### PR TITLE
Added String type annotation. 

### PR DESCRIPTION
**Pull request by fcastellacci**
Fixed a bug in the IRBuilder that made some annotations to be lost.
Added IgnoreNegativeZeroBailout flag for integer-only operations .
Minor refactors

### CODE REVIEW

File: lib/Backend/GlobOpt.cpp [hunk 1]
```diff
@@ -10835,7 +10820,10 @@ GlobOpt::TypeSpecializeIntUnary(
     IR::Opnd *src1 = instr->GetSrc1();
     this->ToInt32(instr, src1, this->currentBlock, src1ValueToSpecialize, nullptr, lossy);
 
-    if(bailOutKind != IR::BailOutInvalid && !this->IsLoopPrePass())
+    bool ignoreOverflowBailout = CONFIG_FLAG(RemoveIntOverflow) && instr->GetSrc1()->GetType() == TyInt32 && bailOutKind == IR::BailOutOnOverflow;
+    bool ignoreNegativeZeroBailouit = CONFIG_FLAG(RemoveNegativeZero) && instr->GetSrc1()->GetType() == TyInt32 && bailOutKind == IR::BailOutOnNegativeZero;
```
Comment (MikeHolman) [inline lib/Backend/GlobOpt.cpp]: nit: ignoreNegativeZeroBailouit -> ignoreNegativeZeroBailout

File: lib/Backend/GlobOpt.cpp [hunk 2]
```diff
@@ -10835,7 +10820,10 @@ GlobOpt::TypeSpecializeIntUnary(
     IR::Opnd *src1 = instr->GetSrc1();
     this->ToInt32(instr, src1, this->currentBlock, src1ValueToSpecialize, nullptr, lossy);
 
-    if(bailOutKind != IR::BailOutInvalid && !this->IsLoopPrePass())
+    bool ignoreOverflowBailout = CONFIG_FLAG(RemoveIntOverflow) && instr->GetSrc1()->GetType() == TyInt32 && bailOutKind == IR::BailOutOnOverflow;
```
Comment (MikeHolman) [inline lib/Backend/GlobOpt.cpp]: this doesn't seem like the right place... seems better for instance to have Instr::ShouldCheckForIntOverflow() check for this flag

**Comment by dilijev** [comment]
@MikeHolman @LouisLaf - Is this pull request intended to be merged to Microsoft:type-annotations? Is any update planned? If this is not intended to be merged, should we close it?

**Comment by dilijev** [comment]
Adding `External Contributor` tag because the author cannot merge this PR on his own and the project maintainers need to determine what to do with this PR.

**Comment by kingalirezamaghsoudi** [comment]
Contents  [hide] 
1	Upgrade guide
2	Summary
3	Main installation guide
4	Alternatives to manual installation
5	Wiki families
6	Appendices
6.1	Advanced uses
6.2	Advanced configuration
7	Installation assistance
7.1	System-specific instructions

**Comment by dilijev** [comment]
@kingalirezamaghsoudi it appears you commented here by mistake. I've removed it. If you feel this action was incorrect, please contact us on Gitter.

**Comment by MikeHolman** [comment]
Closing for now. We can revisit later if we want to integrate any of these changes back to master.

**Pull request closed without merging**